File: src/main/java/com/paragon/client/systems/module/impl/misc/CustomWorld.java
package com.paragon.client.systems.module.impl.misc;

import com.paragon.api.event.network.PacketEvent;
import com.paragon.client.systems.module.Module;
import com.paragon.client.systems.module.Category;
import com.paragon.client.systems.module.setting.Setting;
import me.wolfsurge.cerauno.listener.Listener;
import net.minecraft.network.play.server.SPacketTimeUpdate;

public class CustomWorld extends Module {

    private final Setting<Boolean> customWeather = new Setting<>("Custom Weather", true)
            .setDescription("Set the world weather to a custom value");

    private final Setting<Weather> weather = new Setting<>("Weather", Weather.CLEAR)
            .setDescription("The weather to display")
            .setParentSetting(customWeather);

    private final Setting<Boolean> customTime = new Setting<>("Custom Time", true)
            .setDescription("Set the world time to a custom value");

    private final Setting<Float> time = new Setting<>("Time", 1000f, 0f, 24000f, 1f)
            .setDescription("The time of day")
            .setParentSetting(customTime);

    public CustomWorld() {
        super("CustomWorld", Category.MISC, "Changes the way the world is shown client side");
        this.addSettings(customWeather, customTime);
    }

    @Override
    public void onTick() {
        if (nullCheck()) {
            return;
        }

        mc.world.setRainStrength(weather.getValue().getRainStrength());
        mc.world.setWorldTime(time.getValue().longValue());
    }

    @Listener
    public void onPacketReceive(PacketEvent.PreReceive event) {
        if (event.getPacket() instanceof SPacketTimeUpdate) {
            // Stop the world from updating the time
            event.cancel();
        }
    }

    public enum Weather {
        /**
         * Clear weather - no rain or thunder
         */
        CLEAR(0),

        /**
         * Rainy weather - just rain, darker sky
         */
        RAIN(1),

        /**
         * Thunder - rain + thunder
         */
        THUNDER(2);

        private int rainStrength;

        Weather(int rainStrength) {
            this.rainStrength = rainStrength;
        }

        /**
         * Gets the rain strength
         *
         * @return The rain strength
         */
        public int getRainStrength() {
            return rainStrength;
        }
    }
}

File: src/main/kotlin/com/paragon/impl/module/misc/CustomWorld.kt
package com.paragon.impl.module.misc

import com.paragon.impl.event.network.PacketEvent.PreReceive
import com.paragon.impl.module.Module
import com.paragon.impl.setting.Setting
import com.paragon.bus.listener.Listener
import com.paragon.impl.module.Category
import com.paragon.util.anyNull
import net.minecraft.network.play.server.SPacketTimeUpdate

object CustomWorld : Module("CustomWorld", Category.MISC, "Changes the way the world is shown client side") {

    private val customWeather = Setting(
        "CustomWeather", true
    ) describedBy "Set the world weather to a custom value"

    private val weather = Setting(
        "Weather", Weather.CLEAR
    ) describedBy "The weather to display" subOf customWeather

    private val customTime = Setting(
        "CustomTime", true
    ) describedBy "Set the world time to a custom value"

    private val time = Setting(
        "Time", 1000f, 0f, 24000f, 1f
    ) describedBy "The time of day" subOf customTime

    override fun onTick() {
        if (minecraft.anyNull) {
            return
        }

        minecraft.world.setRainStrength(weather.value.rainStrength.toFloat())
        minecraft.world.worldTime = time.value.toLong()
    }

    @Listener
    fun onPacketReceive(event: PreReceive) {
        if (event.packet is SPacketTimeUpdate) {
            // Stop the world from updating the time
            event.cancel()
        }
    }

    @Suppress("UNUSED")
    enum class Weather(val rainStrength: Int) {
        /**
         * Clear weather - no rain or thunder
         */
        CLEAR(0),

        /**
         * Rainy weather - just rain, darker sky
         */
        RAIN(1),

        /**
         * Thunder - rain + thunder
         */
        THUNDER(2);

    }

}